[DOCS] Add ILM tutorial

mellieA opened the pull request:
## Summary

Added a walkthrough using ILM with filebeat indices

Closes #29363
[skip-ci]

[Preview the tutorial](http://kibana_59502.docs-preview.app.elstc.co/guide/en/kibana/master/example-using-index-lifecycle-policy.html)

elasticmachine commented:
Pinging @elastic/kibana-docs (Team:Docs)

mellieA commented:
This is still a work in progress, I know I need to check on at least the following things: 

* Does a custom policy requires more Filebeat changes
* Add a step to associate a custom policy with the index template

gchaps commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,159 @@
+enabled in the hot phase.
+
+
+** Enable *move to warm phase on rollover*. The index may move to the warm phase
```
move > Move
may > might

gchaps commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,159 @@
+
+
+** Enable *move to warm phase on rollover*. The index may move to the warm phase
+more quickly than intended if it reaches the *Maximum index size* condition before
```
remove the word "condition"

gchaps commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,159 @@
+more quickly than intended if it reaches the *Maximum index size* condition before
+*Maximum age*.
+
+** Provide a value for *Timing for warm phase*. Setting this to *15* will keep the
```
I'd put this first because it is the default.

gchaps commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,159 @@
+indices on hot nodes for a range of  15-45 days, depending on when the initial
+rollover occurred.
+
+* *Select a node attribute to control shard allocation*. For this example,
```
In **Select a node attribute to control shard allocation**, set data:warm(2) to migrate shards to warm data nodes.

gchaps commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,159 @@
+[role="screenshot"]
+image::images/tutorial-ilm-modify-default-warm-phase-rollover.png["Modify to add warm phase"]
+
+* Activate the delete phase.
```
Make this one step.  

Activate the delete phase and set the **Timing** to **60** days.

gchaps commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,159 @@
+To create a custom policy in {kib}, go to *Management > Index Lifecycle Policies > Create Policy*.
+
+
+* *Warm phase*
```
Combine.  

1. Activate the warm phase and configure it as follows.

gchaps commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,159 @@
-* After 60 days, move the index into the cold phase and onto less expensive hardware.
+* *Delete phase*
++
+Activate the delete phase and set the timing.
```
2. Activate the delete phase and set the timing.

mellieA commented:
@gchaps @DanRoscigno I made the changes recommended above and pushed a new commit. 

I'm still waiting to see how my custom policy behaves when it rolls over in 3 days.

mellieA commented:
@gchaps @DanRoscigno I verified that ILM behaves the way I expected with this new policy. I think this is ready for another review!

gchaps commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
 
-* Move the index into the warm phase, shrink the index down to a single shard, 
-and force merge to a single segment.
+. Configure the index to use the new policy in *{kib} > Management > Index Lifecycle
```
This is hard to parse.  Can you make it a series of substeps?
3. Configure the index to use the new policy.
    a. In Index Lifecycle Policies, select your policy.
    b....
    c....

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+
 [[example-using-index-lifecycle-policy]]
-=== Example of using an index lifecycle policy
+=== Tutorial:  Use ILM to manage Filebeat time-based indices
```
We maintain a list of attributes here: https://github.com/elastic/docs/blob/master/shared/attributes.asciidoc

Try to use those when possible. They ensure we can make docs-wide term changes easily.
```suggestion
=== Tutorial:  Use {ilm-cap} to manage {filebeat} time-based indices
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
-=== Example of using an index lifecycle policy
+=== Tutorial:  Use ILM to manage Filebeat time-based indices
+
+With index lifecycle management (ILM), you can create policies to define actions
```
```suggestion
With {ilm} ({ilm-cap}), you can create policies to define actions
```

mellieA replied:
I'm replacing an existing page, so I was leaving the old anchor in case it was referenced elsewhere. Do you still suggest to change it?

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
-=== Example of using an index lifecycle policy
+=== Tutorial:  Use ILM to manage Filebeat time-based indices
+
+With index lifecycle management (ILM), you can create policies to define actions
+that are automatically performed against indices as they age.  In this tutorial,
+you will use {kib}’s *Index Lifecycle Policies* to modify and create ILM policies.
```
> With index lifecycle management (ILM), you can create policies to define actions that are automatically performed against indices as they age.

This implies that ILM only works based on age, but we can also take actions based on index size or number of docs. More importantly, this doesn't really address the _why_ of ILM. It helps you automatically manage your indices for performance and retention. You may want to consider reusing a sentence from https://www.elastic.co/guide/en/elasticsearch/reference/master/overview-index-lifecycle-management.html.

> In this tutorial, you will use Kibana’s Index Lifecycle Policies to modify and create ILM policies.

Nit: I'd reword to avoid future tense if possible. Something like:

```
This tutorial shows you how to use use {kib}'s *Index Lifecycle Policies* to
create and change {ilm-cap} policies.
```

The bolded **Index Lifecycle Policies** seemed unnecessary to me, but maybe that's just a Kibana docs style.

gchaps replied:
Yep, it's a Kibana docs style.

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+you will use {kib}’s *Index Lifecycle Policies* to modify and create ILM policies.
+
+
+[float]
+==== Scenario
```
We should use `[discrete]` instead of `[float]` when possible. I'd also add an explicit anchor here.

```suggestion
[discrete]
[[example-using-index-lifecycle-policy-scenario]]
==== Scenario
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+[float]
+==== Scenario
+
+You’ve been tasked with sending syslog files to an {es} cluster. This
+project has the following data retention guidelines for log data:
```
```suggestion
You’re tasked with sending syslog files to an {es} cluster. This
log data has the following data retention guidelines:
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+You’ve been tasked with sending syslog files to an {es} cluster. This
+project has the following data retention guidelines for log data:
+
+* Keep logs on hot data nodes for 30 days
```
This is the first mention of hot-warm architecture. 

If readers should be familiar with those concepts, maybe add a link to a conceptual doc in the prerequisites?

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+project has the following data retention guidelines for log data:
+
+* Keep logs on hot data nodes for 30 days
+* Rollover to a new index if the size reaches 50GB
```
"Roll over" should be two words when used as a verb. One word as a noun.
```suggestion
* Roll over the index containing log data to a new index when it reaches 50GB
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+* Rollover to a new index if the size reaches 50GB
+* After 30 days:
+** Move the logs to warm data nodes
+** Set replicas to 1
```
> ** Set replicas to 1

I think we need more explanation/context here.

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+* Rollover to a new index if the size reaches 50GB
+* After 30 days:
+** Move the logs to warm data nodes
+** Set replicas to 1
+** Force merge indices to a single segment
```
> ** Set replicas to 1
** Force merge indices to a single segment

I think these two items need more context/explanation. I imagine many users may not be familiar with the concept of replicas, force merges, or segments. At the least, provide a link.

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+* Delete logs after 90 days
+
+
+[float]
+==== Prerequisites
```
Add an explicit anchor here.
```suggestion
[discrete]
[[example-using-index-lifecycle-policy-prerequisites]]
==== Prerequisites
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+To complete this tutorial, you'll need:
+
+* An {es} cluster with hot and warm nodes configured for shard allocation
+awareness. If you’re using https://www.elastic.co/guide/en/cloud/current/ec-getting-started-templates-hot-warm.html[Elasticsearch Service],
```
```suggestion
awareness. If you’re using {cloud}/ec-getting-started-templates-hot-warm.html[Elasticsearch Service],
```

DanRoscigno replied:
I did some more checking into how this works, and I think we should say [`Index-level shard allocation filtering`](https://www.elastic.co/guide/en/elasticsearch/reference/current/shard-allocation-filtering.html) instead of shard allocation awareness.  The Index-level page mentions ILM.

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+choose the hot-warm architecture deployment template.
+
++
+For a self-managed cluster, use https://www.elastic.co/guide/en/elasticsearch/reference/current/allocation-awareness.html[shard allocation awareness]
```
Attributes are especially helpful when linking across books. Generally, avoid using any `/current/` links as they change with each release. They're prone to breaking if a book removes or rearranges its pages.
```suggestion
For a self-managed cluster, use {ref}/allocation-awareness.html[shard allocation awareness]
```

DanRoscigno replied:
This one should point to https://www.elastic.co/guide/en/elasticsearch/reference/current/shard-allocation-filtering.html  @jrodewig would that be `{ref}/shard-allocation-filtering.html[index-level shard allocation filtering]` even though the page is in Elasticsearch docs and not Kibana?

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+node.attr.data: "warm"
+--------------------------------------------------------------------------------
+
+* A server with Filebeat installed and configured to send logs to the `elasticsearch`
```
```suggestion
* A server with {filebeat} installed and configured to send logs to the `elasticsearch`
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+--------------------------------------------------------------------------------
+
+* A server with Filebeat installed and configured to send logs to the `elasticsearch`
+output as described in https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-getting-started.html[Getting Started with Filebeat].
```
```suggestion
output as described in {filebeat-ref}/filebeat-getting-started.html[Getting Started with Filebeat].
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+* A server with Filebeat installed and configured to send logs to the `elasticsearch`
+output as described in https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-getting-started.html[Getting Started with Filebeat].
+
+[float]
+==== View the Filebeat ILM policy
```
Explicit anchor
```suggestion
[discrete]
[[example-using-index-lifecycle-policy-view-fb-ilm-policy]]
==== View the {filebeat} {ilm-init} policy
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+[float]
+==== View the Filebeat ILM policy
+
+Filebeat includes a default ILM policy that enables rollover. Index lifecycle
+management is enabled automatically if you’re using the default `filebeat.yml`
+and index template.
```
```suggestion
{filebeat} includes a default {ilm-init} policy that enables rollover. {ilm-cap}
is enabled automatically if you’re using the default `filebeat.yml`
and index template.
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+To view the default policy in {kib}, go to *Management > Index Lifecycle Policies*,
+search for _filebeat_, and choose the _filebeat-version_ policy.
+
+This policy initiates the rollover action when the index size reaches 50G or 30
+days old.
```
```suggestion
This policy initiates the rollover action when the index size reaches 50GB or 
the index becomes 30 days old.
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
++
+. Set either of the following options to control when the index moves to the warm phase:
+
+** Provide a value for *Timing for warm phase*. Setting this to *15* will keep the
```
```suggestion
** Provide a value for *Timing for warm phase*. Setting this to *15* keeps the
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+
+** Enable *Move to warm phase on rollover*. The index might move to the warm phase
+more quickly than intended if it reaches the *Maximum index size* before
+*Maximum age*.
```
```suggestion
the *Maximum age*.
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+more quickly than intended if it reaches the *Maximum index size* before
+*Maximum age*.
+
+. In *Select a node attribute to control shard allocation*, select *data:warm(2)* in
+the dropdown to migrate shards to warm data nodes.
```
```suggestion
. In the *Select a node attribute to control shard allocation* dropdown, select *data:warm(2)* to
migrate shards to warm data nodes.
```

DanRoscigno commented:
I have an [Elasticsearch docs PR](https://github.com/elastic/elasticsearch/pull/54592) that this PR depends on.  It is waiting on a review.

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+[float]
+==== Create a custom policy
+
+If meeting a specific retention time period is your most important priority, you
+can create a custom policy.  For this option, you will use Filebeat daily indices
+without rollover.
```
```suggestion
If meeting a specific retention time period important, you
can create a custom policy. For this option, you will use {filebeat} daily indices
without rollover.
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
 
-* After 60 days, move the index into the cold phase and onto less expensive hardware.
++
+NOTE: If you initially used the default Filebeat ILM policy, you will see a notice that the
```
```suggestion
NOTE: If you initially used the default {filebeat} {ilm-init} policy, you will see a notice that the
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+TIP: When you change the policy associated with the index template, the active
+index will continue to use the policy it was associated with at index creation
+unless you manually update it. The next new index will use the updated policy.
+For more reasons that your ILM policy changes might be delayed, see https://www.elastic.co/guide/en/elasticsearch/reference/current/update-lifecycle-policy.html#update-lifecycle-policy[Update Lifecycle Policy].
```
```suggestion
For more reasons that your {ilm-init} policy changes might be delayed, see {ref}/update-lifecycle-policy.html#update-lifecycle-policy[Update Lifecycle Policy].
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
-and force merge to a single segment.
+. Configure the index to use the new policy in *{kib} > Management > Index Lifecycle
+Policies > _Select your custom policy_ > Actions > Add policy to template > Select
+your Filebeat index template*
```
```suggestion
your Filebeat index template*.
```

DanRoscigno commented:
@elasticmachine run elasticsearch-ci/docs

DanRoscigno commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,167 @@
+
+To complete this tutorial, you'll need:
+
+* An {es} cluster with hot and warm nodes configured for shard allocation
+awareness. If you’re using https://www.elastic.co/guide/en/cloud/current/ec-getting-started-templates-hot-warm.html[Elasticsearch Service],
```
I did some more checking and index level allocation appears to be what is used, it is related to shard allocation awareness. https://www.elastic.co/guide/en/elasticsearch/reference/current/shard-allocation-filtering.html

  The Index-level page mentions ILM.
```suggestion
* An {es} cluster with hot and warm nodes configured for index-level shard allocation
filtering. If you’re using https://www.elastic.co/guide/en/cloud/current/ec-getting-started-templates-hot-warm.html[Elasticsearch Service],
```

mellieA replied:
Updated to add shard allocation filtering.

DanRoscigno commented:
I think it looks great @mellieA , I made a suggestion to change the shard awareness page.

mellieA commented:
Thank you @gchaps, @DanRoscigno, and @jrodewig for all of the suggestions.

* Updated to use attributes
* Updated links to use references instead of full version specific URLs
* Added context for replicas and forcemerge
* Revised the introduction to link to the Elasticsearch ILM overview page
* Added anchors and updated to use `[discrete]` instead of `[float]`
* Incorporated other small edits

Please take a look and let me know if I missed anything!

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,177 @@
+
 [[example-using-index-lifecycle-policy]]
-=== Example of using an index lifecycle policy
+=== Tutorial:  Use {ilm-init} to manage {filebeat} time-based indices
```
```suggestion
=== Tutorial: Use {ilm-init} to manage {filebeat} time-based indices
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,177 @@
-=== Example of using an index lifecycle policy
+=== Tutorial:  Use {ilm-init} to manage {filebeat} time-based indices
+
+With {ilm} ({ilm-init}), you can create policies perform actions automatically 
```
```suggestion
With {ilm} ({ilm-init}), you can create policies that perform actions automatically 
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,177 @@
+=== Tutorial:  Use {ilm-init} to manage {filebeat} time-based indices
+
+With {ilm} ({ilm-init}), you can create policies perform actions automatically 
+against indices as they age and grow. {ilm-init} policies help you to manage
```
```suggestion
on indices as they age and grow. {ilm-init} policies help you to manage
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,177 @@
+
+With {ilm} ({ilm-init}), you can create policies perform actions automatically 
+against indices as they age and grow. {ilm-init} policies help you to manage
+performance, resilience, and retention of your data during its lifecycle. This tutorial will show
```
```suggestion
performance, resilience, and retention of your data during its lifecycle. This tutorial shows
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,177 @@
+performance, resilience, and retention of your data during its lifecycle. This tutorial will show
+you how to use {kib}’s *Index Lifecycle Policies* to modify and create {ilm-init}
+policies. You can learn more about all of the actions, benefits, and lifecycle
+phases in the {ref}/overview-index-lifecycle-management.html[{ilm} overview].
```
```suggestion
phases in the {ref}/overview-index-lifecycle-management.html[{ilm-init} overview].
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,177 @@
+[[example-using-index-lifecycle-policy-view-fb-ilm-policy]]
+==== View the {filebeat} {ilm-init} policy
+
+{filebeat} includes a default {ilm-init} policy that enables rollover. {ilm-cap}
```
Not sure why we spell out ILM here.
```suggestion
{filebeat} includes a default {ilm-init} policy that enables rollover. {ilm-init}
```

jrodewig commented on docs/management/index-lifecycle-policies/example-index-lifecycle-policy.asciidoc:
```diff
@@ -1,23 +1,177 @@
+NOTE: If you initially used the default {filebeat} {ilm-init} policy, you will
+see a notice that the template already has a policy associated with it. Confirm
+that you want to overwrite that configuration.
 
```
The `NOTE` admonition above is inline with the previous numbered item. The `TIP` admonition is not.

It's fine either way. Just wanted to point that out.

If you'd like both to be inline, just add a `+` to line 172.

gchaps commented:
@elasticmachine merge upstream
